### ISSUE TITLE
Proposal: Remove logit softcap and rely solely on QK-norm to eliminate redundant computational overhead

### ISSUE THREAD

**Issue by HosLak** [karpathy/nanochat#362]
In the current implementation of nanochat, both logit softcap (e.g., applying a capping function like tanh to attention logits) and QK-norm (normalizing queries and keys before dot-product) are used simultaneously in the attention mechanism. While both techniques aim to control attention logit growth and prevent training instabilities, combining them introduces unnecessary computational overhead without proportional benefits. This can increase latency during training and inference, especially in resource-constrained environments, and complicates the codebase without improving stability or performance.

**Proposed Change:**

I suggest removing the logit softcap entirely and relying exclusively on QK-norm. QK-norm addresses the root cause of logit explosions by normalizing queries (Q) and keys (K) upfront, preserving attention entropy and enabling stable training at scale. Softcap, on the other hand, is a post-processing fix that clips logits after computation, which can be redundant when QK-norm is already in place. This change would simplify the architecture, reduce ops (e.g., no extra tanh or clipping per attention head), and align with recent advancements in Transformer scaling.

**Rationale and Evidence:**

- **Redundancy and Overhead**: QK-norm proactively prevents uncontrolled logit growth (e.g., values exceeding 50,000), avoiding one-hot attention weights and gradient divergences. Softcap adds an extra layer of processing that may over-constrain the model, potentially reducing diversity in attention distributions or requiring additional hyperparameter tuning. Experiments in related works show that QK-norm alone suffices for stability, making the combination inefficient—similar to adding redundant safeguards that increase compute without gains.

- **References to Relevant Papers:**

  - **Gemma 3 Technical Report (Google DeepMind, 2024)** [arXiv:2503.19786](https://arxiv.org/abs/2503.19786): In Gemma 3, the soft-capping from Gemma 2 is explicitly replaced with QK-norm, inspired by Dehghani et al. (2023), Wortsman et al. (2023), and Chameleon (2024). This shift improves efficiency for long-context (up to 128K tokens) and multimodal handling, implying that QK-norm is a more comprehensive solution for stability without the need for post-hoc capping.

  - **Scaling Vision Transformers to 22 Billion Parameters (Dehghani et al., 2023)** [arXiv:2302.05442](https://arxiv.org/abs/2302.05442): Introduces QK-norm as essential for scaling ViTs to 22B parameters. It stabilizes training by preventing logit growth and entropy collapse, allowing higher learning rates (e.g., 1e-3) and avoiding divergences seen in unnormalized models. The paper emphasizes QK-norm's role in efficient hardware utilization (e.g., 54.9% MFU on TPUv4) and scaling benefits like better OOD robustness, presenting it as superior to alternatives like reduced learning rates—no mention of needing soft-capping alongside it.

  - **Chameleon: Mixed-Modal Early-Fusion Foundation Models (Chameleon Team, 2024)** [arXiv:2405.09818](https://arxiv.org/abs/2405.09818): Uses QK-norm to prevent "logit drift" in multimodal early-fusion models, where competing modalities cause norm escalation and training divergence. Ablations show that without QK-norm, models like Chameleon-7B diverge mid-training, but with it, stable pre-training on 9.2T tokens is achieved. It's combined with z-loss but not soft-capping, highlighting QK-norm's sufficiency for long-context and scaled multimodal stability.

  - **Small-scale proxies for large-scale Transformer training instabilities (Wortsman et al., 2023)** [arXiv:2309.17421](https://arxiv.org/abs/2309.17421): While not directly detailing QK-norm vs. softcap, it discusses proxies for instabilities like attention collapses and recommends normalization techniques (inspired by this, Gemma 3 adopts QK-norm). It shows that such methods reduce sensitivity to learning rates by up to 3x, enabling scalable training without additional patches.

This proposal is based on these empirical findings, where QK-norm consistently emerges as a standalone, efficient stabilizer. If adopted, it could be tested via ablation studies on nanochat's training runs to confirm no regression in performance.

**Comment by svlandeg** [comment]
Hi! Have you already done any experiments to test this?

**Comment by HosLak** [comment]
> Hi! Have you already done any experiments to test this?

Hi! No, I don't have the hardware to run experiments myself, so I relied on references from existing papers to support the proposal. The image below is from another relevant paper I hadn't mentioned before (arXiv:2410.16682):

<img width="708" height="128" alt="Image" src="https://github.com/user-attachments/assets/08f9a135-8c59-41e5-84fe-1fcc96ee0961" />

which shows perplexity results for various methods, including the combination of QK-norm with soft capping (QK norm cap at 11.00 perplexity). Notably, the paper discusses how this combo provides additional stability benefits, such as allowing 1.5x higher learning rates (up to 80e-3 without divergence) compared to QK-norm alone (up to 60e-3). While standalone QK-norm has slightly better perplexity (10.84), the combination improves over the baseline and could be worth evaluating in nanochat for more aggressive training.

**Comment by casinca** [comment]
Although my opinion is meaningless, I can vouch for what @HosLak said, it's really common in recent OSS release to only have QKNorm, I haven't seen logit softcapping since Gemma2.

You can add Qwen3 and Olmo 3 in the QKNorm camp too. I haven't run the code here so I can't claim it's going to be a thing but just that it's a quite common choice at the moment.

**Comment by karpathy** [comment]
I'll just say that modded-nanogpt still has it, and also if I naively remove it the results are worse.

**Comment by karpathy** [comment]
Closing. If someone can get better results I'll merge.

**Comment by ediestel** [comment]
Issue #362: Proposal to remove logit softcap:

resolved #666 under the follwing constraints; not sure how this scales.
                                                                              
  - d4 model (tiny), not d24 (GPT-2 scale)
  - 200 steps with AdamW, not thousands of steps with Muon                                                                           
  - Max logit reached 0.19 — but Karpathy's d24 with Muon may push logits much higher
